src/anyp/Uri.cc

```diff
@@ -1008,56 +1008,54 @@ AnyP::Uri::Uri(AnyP::UriScheme const &aScheme) :
 }
 
 // TODO: fix code duplication with AnyP::Uri::parse()
-char *
-AnyP::Uri::cleanup(const char *uri)
+SBuf
+AnyP::Uri::Cleanup(const SBuf &uri)
 {
-    char *cleanedUri = nullptr;
+    // 'whitespace' for chop and strip are different
+    static const CharacterSet wspChop = (CharacterSet::WSP +
+                                         CharacterSet::LF +
+                                         CharacterSet::CR).rename("wsp-chop");
+#if 0 // XXX: use these instead of xisspace() character match (it varies)
+    static const CharacterSet wspStrip = (CharacterSet("VT,FF","\v\f") +
+                                          wspChop).rename("wsp-strip");
+#endif
+
+    int flags = 0;
+    SBuf cleanedUri;
     switch (Config.uri_whitespace) {
     case URI_WHITESPACE_ALLOW: {
-        const auto flags = RFC1738_ESCAPE_NOSPACE | RFC1738_ESCAPE_UNESCAPED;
-        cleanedUri = xstrndup(rfc1738_do_escape(uri, flags), MAX_URL);
-        break;
+        flags |= RFC1738_ESCAPE_NOSPACE | RFC1738_ESCAPE_UNESCAPED;
+        SBuf tmp = uri;
+        cleanedUri = SBuf(rfc1738_do_escape(tmp.c_str(), flags));
```

Comment:
Why not Encode?

Reply:
FWIW, I am removing this `allow` feature in the followup parser changes. So do not consider it worth improving right now.